### ISSUE TITLE
NativeAOT: Failed to compile virtual statics with default implementation

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#70930]
### Description

Repro:
```csharp
using System.Numerics;

static U Foo<T, U>(U x, U y)
    where T : I1, I2
    where U : IAdditionOperators<U, U, U>,
        ISubtractionOperators<U, U, U>,
        IMultiplyOperators<U, U, U>,
        IDivisionOperators<U, U, U>
    => Foo1<T, U>(x, y) + Foo2<T, U>(x, y);
static U Foo1<T, U>(U x, U y)
    where T : I1
    where U : IAdditionOperators<U, U, U>,
        ISubtractionOperators<U, U, U>,
        IMultiplyOperators<U, U, U>,
        IDivisionOperators<U, U, U>
    => T.Foo(x, y);
static U Foo2<T, U>(U x, U y)
    where T : I2
    where U : IAdditionOperators<U, U, U>,
        ISubtractionOperators<U, U, U>,
        IMultiplyOperators<U, U, U>,
        IDivisionOperators<U, U, U>
    => T.Foo(x, y);

Console.WriteLine(Test());

static int Test()
{
    int a = default;
    int b = default;
    int sum = default;
    sum += Foo<C1, int>(++a, ++a);
    sum *= Foo<C2, int>(++b, ++b);
    a = sum * 2;
    sum /= Foo<C3, int>(++a * b, a / b);
    sum -= a + b;
    sum -= Foo1<C4, int>(--a, ++b);
    b = sum / 2;
    sum += Foo2<C5, int>(a / b, a * b - 1);
    sum *= Foo<C6, int>(a, b);
    return sum;
}

interface I1
{
    virtual static T Foo<T>(T l, T r)
        where T : IAdditionOperators<T, T, T>,
            ISubtractionOperators<T, T, T>,
            IMultiplyOperators<T, T, T>,
            IDivisionOperators<T, T, T>
        => l + r;
}
interface I2
{
    abstract static T Foo<T>(T l, T r)
        where T : IAdditionOperators<T, T, T>,
            ISubtractionOperators<T, T, T>,
            IMultiplyOperators<T, T, T>,
            IDivisionOperators<T, T, T>;
}
interface I : I1, I2
{
    static T I1.Foo<T>(T l, T r) => l - r;
    static T I2.Foo<T>(T l, T r) => l * r;
}
class C1 : I { }
class C2 : I
{
    public static T Foo<T>(T l, T r)
        where T : IAdditionOperators<T, T, T>,
            ISubtractionOperators<T, T, T>,
            IMultiplyOperators<T, T, T>,
            IDivisionOperators<T, T, T>
        => l / r;
}
class C3 : I
{
    static T I1.Foo<T>(T l, T r) => l + r * l;
    static T I2.Foo<T>(T l, T r) => l - r * l;
}
class C4 : I1 { }
class C5 : I2
{
    public static T Foo<T>(T l, T r)
        where T : IAdditionOperators<T, T, T>,
            ISubtractionOperators<T, T, T>,
            IMultiplyOperators<T, T, T>,
            IDivisionOperators<T, T, T>
        => l / r - r;
}
class C6 : I1, I2
{
    public static T Foo<T>(T l, T r)
        where T : IAdditionOperators<T, T, T>,
            ISubtractionOperators<T, T, T>,
            IMultiplyOperators<T, T, T>,
            IDivisionOperators<T, T, T>
        => l / r + r;
}
```

### Reproduction Steps

Build with NativeAOT

### Expected behavior

Produce a succesful build.

### Actual behavior

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at ILCompiler.DependencyAnalysis.ConstrainedMethodUseLookupResult.GetTarget(NodeFactory factory, GenericLookupResu
ltContext dictionary) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/GenericLookupResult
.cs:line 1459
   at ILCompiler.DependencyAnalysis.ReadyToRunGenericHelperNode.InstantiateDependencies(NodeFactory factory, Instanti
ation typeInstantiation, Instantiation methodInstantiation) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/
DependencyAnalysis/ReadyToRunGenericHelperNode.cs:line 167
   at ILCompiler.DependencyAnalysis.ShadowConcreteMethodNode.GetStaticDependencies(NodeFactory factory) in /_/src/cor
eclr/tools/Common/Compiler/DependencyAnalysis/ShadowConcreteMethodNode.cs:line 80
 at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node
) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 181
   at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1 node) in
 /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 221
   at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() in /_/src/coreclr/tools/aot/ILCo
mpiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 256
   at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/IL
Compiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 307
   at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILSc
anner.cs:line 140
   at ILCompiler.Program.Run(String[] args) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 833
   at ILCompiler.Program.Main(String[] args) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 1090
```

**Comment by am11** [comment]
Library free repro:
```sh
$ dotnet new console -n generic1; cd generic1
$ cat > Program.cs << EOF
Bar<C>();

static void Bar<T>() where T : J => T.Foo();

interface I
{
    abstract static void Foo();
}

interface J : I
{
    static void I.Foo() { }
}

class C : J { }
EOF

$ dotnet publish --use-current-runtime -p:PublishAot=true -p:LangVersion=preview

EXEC : error : Object reference not set to an instance of an object. [/home/am11/projects/generic1/generic1.csproj]
  System.NullReferenceException: Object reference not set to an instance of an object.
     at ILCompiler.DependencyAnalysis.ConstrainedMethodUseLookupResult.GetTarget(NodeFactory factory, GenericLookupResultContext dictionary) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/GenericLookupResult.cs:line 1459
     at ILCompiler.DependencyAnalysis.ReadyToRunGenericHelperNode.InstantiateDependencies(NodeFactory factory, Instantiation typeInstantiation, Instantiation methodInstantiation) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunGenericHelperNode.cs:line 167
     at ILCompiler.DependencyAnalysis.ShadowConcreteMethodNode.GetStaticDependencies(NodeFactory factory) in /_/src/coreclr/tools/Common/Compiler/DependencyAnalysis/ShadowConcreteMethodNode.cs:line 80
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 181
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 319
     at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in /_/src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilation.cs:line 88
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:line 526
     at ILCompiler.Program.Run(String[] args) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 893
     at ILCompiler.Program.Main(String[] args) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 1085
/home/am11/.dotnet7/sdk/7.0.100-preview.6.22314.18/Sdks/Microsoft.DotNet.ILCompiler/build/Microsoft.NETCore.Native.targets(274,5): error MSB3073: The command ""/home/am11/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0-preview.6.22312.1/tools/ilc" @"obj/Debug/net7.0/linux-x64/native/generic1.ilc.rsp"" exited with code 1. [/home/am11/projects/generic1/generic1.csproj]
```

**Comment by hez2010** [comment]
Note that the library free repro doesn't cover diamond shape. I'm not sure whether it is also causing any issue.

**Comment by MichalStrehovsky** [comment]
Not surprised this doesn't work, because it's not implemented.

**Comment by MichalStrehovsky** [comment]
Thank you for the fun test cases. Fix is in #72661.
